Internal error when specialization for TypeVarTuple contains multiple unpacked tuples

The following code results in a mypy "INTERNAL ERROR".

```python
class A[*Ts]: ...

A[*tuple[int, ...], *tuple[int, ...]]  # no error, fails at runtime:
# TypeError: More than one unpacked arbitrary-length tuple argument

b: tuple[*tuple[int, ...], *tuple[int, ...]]  # error, as expected
```

Here's the stack trace for the crash:

```
version: 1.15.0
Traceback (most recent call last):
  File "mypy/checkexpr.py", line 5903, in accept
  File "mypy/nodes.py", line 2488, in accept
  File "mypy/checkexpr.py", line 4768, in visit_type_application
  File "mypy/checkexpr.py", line 4963, in apply_type_arguments_to_callable
  File "mypy/checkexpr.py", line 4899, in split_for_callable
  File "mypy/typeanal.py", line 2532, in validate_instance
  File "mypy/types.py", line 3732, in find_unpack_in_list
AssertionError: 
```

I found this when diagnosing a [bug reported against pyright](https://github.com/microsoft/pyright/issues/10219).

It may be related to https://github.com/python/mypy/issues/17755.

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -6071,9 +6071,8 @@ class SemanticAnalyzer(
                 return None
             types.append(analyzed)
 
-        analyzer = self.type_analyzer()
-        types = analyzer.check_unpacks_in_list(types)
-
+        if allow_unpack:
+            types = self.type_analyzer().check_unpacks_in_list(types)
         if has_param_spec and num_args == 1 and types:
             first_arg = get_proper_type(types[0])
             single_any = len(types) == 1 and isinstance(first_arg, AnyType)

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -6071,6 +6071,9 @@ class SemanticAnalyzer(
                 return None
             types.append(analyzed)
 
+        analyzer = self.type_analyzer()
+        types = analyzer.check_unpacks_in_list(types)
+
         if has_param_spec and num_args == 1 and types:
             first_arg = get_proper_type(types[0])
             single_any = len(types) == 1 and isinstance(first_arg, AnyType)

--- a/mypy/typeanal.py
+++ b/mypy/typeanal.py
@@ -2006,7 +2006,7 @@ class TypeAnalyser(SyntheticTypeVisitor[Type], TypeAnalyzerPluginInterface):
 
         if num_unpacks > 1:
             assert final_unpack is not None
-            self.fail("More than one Unpack in a type is not allowed", final_unpack)
+            self.fail("More than one Unpack in a type is not allowed", final_unpack.type)
         return new_items
 
     def tuple_type(self, items: list[Type], line: int, column: int) -> TupleType:

--- a/test-data/unit/check-python312.test
+++ b/test-data/unit/check-python312.test
@@ -2035,7 +2035,10 @@ class Z: ...  # E: Name "Z" already defined on line 2
 class A[*Ts]: ...
 
 A[*tuple[int, ...], *tuple[int, ...]]  # E: More than one Unpack in a type is not allowed
+a: A[*tuple[int, ...], *tuple[int, ...]]  # E: More than one Unpack in a type is not allowed
+def foo(a: A[*tuple[int, ...], *tuple[int, ...]]): ...  # E: More than one Unpack in a type is not allowed
 
+tuple[*tuple[int, ...], *tuple[int, ...]]  # E: More than one Unpack in a type is not allowed
 b: tuple[*tuple[int, ...], *tuple[int, ...]]  # E: More than one Unpack in a type is not allowed
 [builtins fixtures/tuple.pyi]
 [typing fixtures/typing-full.pyi]

--- a/test-data/unit/check-python312.test
+++ b/test-data/unit/check-python312.test
@@ -2029,3 +2029,13 @@ def foo() -> None:
 class Z: ...  # E: Name "Z" already defined on line 2
 [builtins fixtures/tuple.pyi]
 [typing fixtures/typing-full.pyi]
+
+[case testPEP695MultipleUnpacksInBareApplicationNoCrash]
+# https://github.com/python/mypy/issues/18856
+class A[*Ts]: ...
+
+A[*tuple[int, ...], *tuple[int, ...]]  # E: More than one Unpack in a type is not allowed
+
+b: tuple[*tuple[int, ...], *tuple[int, ...]]  # E: More than one Unpack in a type is not allowed
+[builtins fixtures/tuple.pyi]
+[typing fixtures/typing-full.pyi]

